### DISCUSSION TITLE
Calibre 5.42 + DeDrm 7.2.1: not working on Mac OS with AZW files

### DISCUSSION THREAD

**Discussion by cstadler** [apprenticeharper/DeDRM_tools#2064] [Q&A] [answered]
**Observed behavior:**
1. installed Calibre 5.42 on Max OS Monterey (12.3.1)
2. installed DeDRM plugin 7.2.1
3. restarted Calibre
4. dragged AZW files into Calibre
5. double click on AZQ file to open in desktop reader
6. DRM dialogue appears

This is my normal flow for verifying DeDRM success. It was working as recently as last week. Any suggestions?

Here is my debug log:

calibre Debug log
calibre 5.42  embedded-python: True is64bit: True
macOS-10.16-x86_64-i386-64bit Darwin ('64bit', '')
('Darwin', '21.4.0', 'Darwin Kernel Version 21.4.0: Fri Mar 18 00:45:05 PDT 2022; root:xnu-8020.101.4~15/RELEASE_X86_64')
Python 3.8.5
OSX: ('10.16', ('', '', ''), 'x86_64')
Interface language: None
Successfully initialized third party plugins: DeDRM (7, 2, 1) && Obok DeDRM (7, 2, 1)
calibre 5.42  embedded-python: True is64bit: True
macOS-10.16-x86_64-i386-64bit Darwin ('64bit', '')
('Darwin', '21.4.0', 'Darwin Kernel Version 21.4.0: Fri Mar 18 00:45:05 PDT 2022; root:xnu-8020.101.4~15/RELEASE_X86_64')
Python 3.8.5
OSX: ('10.16', ('', '', ''), 'x86_64')
Interface language: None
Successfully initialized third party plugins: DeDRM (7, 2, 1) && Obok DeDRM (7, 2, 1)
devicePixelRatio: 2.0
logicalDpi: 72.0 x 72.0
physicalDpi: 147.50000221522777 x 147.0000022077185
Using calibre Qt style: True
[0.00] Starting up...
[0.05] Showing splash screen...
[0.24] splash screen shown
[0.24] Initializing db...
[0.25] db initialized
[0.25] Constructing main UI...
DEBUG:    0.0 obok::utilities.py - loading translations
DEBUG:    0.0 obok::dialogs.py - loading translations
DEBUG:    0.0 obok::config.py - loading translations
DEBUG:    0.0 obok::action_err.py - loading translations
[1.00] main UI initialized...
[1.00] Hiding splash screen
Starting QuickView
DeDRM v7.2.1: Trying to decrypt The Unbroken_ 1 (Magic of the Lost).azw3
Using Library AlfCrypto Python
Using Library AlfCrypto Python
MobiDeDrm v1.0.
Copyright © 2008-2020 The Dark Reverser, Apprentice Harper et al.
Decrypting Kindle Format 8 ebook: The Unbroken: 1 (Magic of the Lost)
Found 2 keys to try after 0.1 seconds
Crypto Type is: 2
DeDRM v7.2.1: Failed to decrypt with error: No key found in 2 PIDs tried.
DeDRM v7.2.1: Looking for new default Kindle Key after 0.1 seconds
Found k4Mac kinf2018 file: /Users/christian/Library/Application Support/Kindle/storage/.kinf2018
trying username  b'christian'  on file  /Users/christian/Library/Application Support/Kindle/storage/.kinf2018
trying IDString: b'55bd3d5f289d'
Decrypted key file using IDString '55bd3d5f289d' and UserName 'christian'
DeDRM v7.2.1: Found 1 new key
MobiDeDrm v1.0.
Copyright © 2008-2020 The Dark Reverser, Apprentice Harper et al.
Decrypting Kindle Format 8 ebook: The Unbroken: 1 (Magic of the Lost)
Got SerialNumber from database default_key_1
Got UsernameHash from database default_key_1
Found 4 keys to try after 0.4 seconds
Crypto Type is: 2
DeDRM v7.2.1: Ultimately failed to decrypt after 0.4 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 540, in KindleMobiDecrypt
    book = k4mobidedrm.GetDecryptedBook(path_to_ebook,kindleDatabases,androidFiles,serials,pids,self.starttime)
  File "calibre_plugins.dedrm.k4mobidedrm", line 238, in GetDecryptedBook
    mb.processBook(totalpids)
  File "calibre_plugins.dedrm.mobidedrm", line 484, in processBook
    raise DrmException("No key found in {0:d} PIDs tried.".format(len(goodpids)))
calibre_plugins.dedrm.mobidedrm.DrmException: No key found in 2 PIDs tried.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre/customize/ui.py", line 175, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 633, in run
    decrypted_ebook = self.KindleMobiDecrypt(path_to_ebook)
  File "calibre_plugins.dedrm.__init__", line 583, in KindleMobiDecrypt
    raise DeDRMError("{0} v{1}: Ultimately failed to decrypt after {2:.1f} seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md".format(PLUGIN_NAME, PLUGIN_VERSION,time.time()-self.starttime))
calibre_plugins.dedrm.DeDRMError: DeDRM v7.2.1: Ultimately failed to decrypt after 0.4 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
[14.52] splash screen hidden
[14.52] Started up in 14.52 seconds with 48 books
Added The Unbroken: 1 (Magic of the Lost) to db in: 0.0
Added 1 books in 0.8 seconds
Unable to open device <calibre.devices.kindle.driver.KINDLE2 object at 0x10af232e0>
Traceback (most recent call last):
  File "calibre/devices/usbms/device.py", line 718, in open
  File "calibre/devices/usbms/device.py", line 393, in open_osx
  File "calibre/devices/usbms/device.py", line 382, in osx_bsd_names
  File "calibre/devices/usbms/device.py", line 341, in _osx_bsd_names
calibre.devices.errors.DeviceError: Could not detect BSD names for Kindle 2/3/4/Touch/PaperWhite/Voyage Device Interface. Try rebooting.
Output from osx_get_usb_drives():
[]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre/gui2/device.py", line 233, in do_connect
  File "calibre/devices/usbms/device.py", line 721, in open
  File "calibre/devices/usbms/device.py", line 393, in open_osx
  File "calibre/devices/usbms/device.py", line 382, in osx_bsd_names
  File "calibre/devices/usbms/device.py", line 341, in _osx_bsd_names
calibre.devices.errors.DeviceError: Could not detect BSD names for Kindle 2/3/4/Touch/PaperWhite/Voyage Device Interface. Try rebooting.
Output from osx_get_usb_drives():
[]

Connect to device failed, retrying in 5 seconds...
Worker Launch took: 0.01 seconds

**Comment by ElleKayEm** [comment, marked as answer]
That file name and azw3 extension looks like a book downloaded from the website.  Did you choose an E-ink Kindle (not a Fire) for the device?  Do you have that Kindle's serial number entered correctly into DeDRM's customization?

**Reply by cstadler** [reply]
Thank you! I had to add my serial number for my new reader. I had added some other books and they were working, but I think they may have been using my old device's serial number (which was deprecated when I de-registered it?). In any case, thank you for the nudge in the right direction.
